### ISSUE TITLE
Manually sleeping/waking `Watchexec`

### ISSUE THREAD

**Issue by emilHof** [watchexec/watchexec#579]
Hey, apologies in advance for the maybe slightly silly question!

There is a problem I am facing with having multiple `Watchexec`s watching binaries in the same crate. Right now it is sometimes the case that two watchers get triggered at the same time, and both having had changes to their source code causes them to recompile simultaneously. It can then happen that one `Watchexec` ends up holding a lock on the compilation process but is unable to finish due to the other `Watchexec`(s) also holding a lock. I.e. there ends up being a rather saddening deadlock in the compilation process...

I was curious if you guys know of a nice and easy way of getting around this sort of issue, by maybe putting watchers to sleep/preventing them from rerunning their command, unless explicitly told to so?

If this is not really a thing that you can do, as it seems to be slightly out of the scope of the lightweight and clean tool, `watchexec`, then I also totally understand!

Thank you so much for maintaining such a nice package! :))

**Comment by passcod** [comment]
I would say: don't do that! ;)

I think this is a situation where you kinda need to turn the issue around and make a little tool to manage your processes under a single watchexec instance. You can use the `--emit-events-to` option introduced last release to get what data you need to figure out which processes should be managed how.

If you do want to use watchexec for its process management, one pattern that has worked for others is to spin up watchexec instances that watch a single file, and then `touch` that file from the main watchexec whenever you want to restart that process.

**Comment by emilHof** [comment]
Hey @passcod, thank you so much for the great advice!

I am not quite certain how processes under watchexec operate, but i will look at that and see what i can learn! by default, is there usually one process under the watchexec instance?

For a bit more context, the project is using the crate version of `watchexec`, but it simply spanws a multitude of `Watchexec` instances, one for each function. Is there a way to extract the events from a `Watchexec` and feed it to the processes, or is the second part of your response referring to how to do this? this has me a little confused, i am sorry haha!

thank you for your help with this ! :))

**Comment by passcod** [comment]
Ah, the library version. Yeah, so, what the CLI does is use the `pre_spawn` hook, which has access to the events, and then it sets an environment variable on the process. What I'd do in your case is instead of having one instance per function which watches the filesystem is **either**:
1. probably easier but a bit hacky
  - have one single instance which watches the filesystem, with a command to a program which does nothing.
  - in the `pre_spawn` hook, use the events data to figure out which function-specific instance(s) should be pinged, and send message(s) along a channel to that effect
  - function-specific instances don't watch anything, but instead you watch a channel (as above) for a particular message and when one arrives you [insert a synthetic event into the instance's event queue](https://github.com/watchexec/watchexec/blob/main/crates/cli/src/lib.rs#L119), which will kick off the usual.

2. more boilerplate but more efficient
  - instead of using a Watchexec instance, reconstruct what it's doing. look into the source for this, but you want:
  - [the event queue](https://github.com/watchexec/watchexec/blob/main/crates/lib/src/watchexec.rs#L79)
  - [the fs worker](https://github.com/watchexec/watchexec/blob/main/crates/lib/src/watchexec.rs#L112)
  - [more than one action workers](https://github.com/watchexec/watchexec/blob/main/crates/lib/src/watchexec.rs#L108-L111), one for each function
  - you'll need to figure out how to distribute events to multiple action workers
  - each action worker will essentially be a lightweight Watchexec instance, with its own process and state and so on

**Comment by emilHof** [comment]
Wow, @passcod, thank you for writing up such a detailed sponse!

That makes sense; right now the project i am working on is also using `pre_spawn` to do a couple of its setup steps, but being honest, the code base is not at all written by me and i am still trying to get an understanding of how `Watchexec` is being used.

Option (1) does seem a lot more straightforward, you are right! Does that mean that `pre_spawn` is fired every time an event is received by the `Wachexec` instance? One of the major difficulties, with option (1) or option (2) would probably be figuring out which binary watcher the event should be routed to, as the tool is supposed to support multiple binaries, but also cargo workspaces, yet i believe you are pointing that out as well..

Option (2) would be really nice, especially as it would probably end up being a lot cleaner and integrating much more nicely! And, is you are pointing out, a lot more efficient! Does `watchexec` also expose the `fs worker`s and `action worker`s?

**Comment by passcod** [comment]
`pre_spawn` is fired just before watchexec would spawn (hence the name) a command. Now that I think about it you could instead use the `action` hook and always return an `Outcome::DoNothing`.

The lifecycle of watchexec goes like... I'm going to quickly draw something here, should really get something a little more official in the docs.

<img width="814" alt="paint" src="https://user-images.githubusercontent.com/155787/236479361-8729fd10-f4a4-4193-87ec-6cd877016d47.png">

So, events sourced from the file system, processes launched by watchexec, stdin, signals, and events manually sent in via `Watchexec::send_event()` all go into a channel. (That channel is actually also a priority queue, so that more urgent events like `Ctrl-C` can bypass a congested event queue, for example.)

That channel is consumed by the action worker. The action worker starts by the "throttle + filter" step, which is [a big loop](https://docs.rs/watchexec/latest/src/watchexec/action/worker.rs.html#40-120) that builds up a Vec of Events that pass [filters](https://docs.rs/watchexec/latest/watchexec/action/struct.WorkingData.html#structfield.filterer) within the "[throttle duration](https://docs.rs/watchexec/latest/watchexec/action/struct.WorkingData.html#structfield.throttle)". Once it's "out of throttle", ie it's reached the end of the throttle duration _and_ it has at least one event that's passed the filters, it makes an [`Action`](https://docs.rs/watchexec/latest/watchexec/action/struct.Action.html) which essentially wraps that `Vec<Event>`.

Then the [`action` handler](https://docs.rs/watchexec/latest/watchexec/action/struct.WorkingData.html#structfield.action_handler) is called, with this `Action` as argument. The handler can do anything (but should return quickly to avoid holding up Watchexec) but the main thing it does is it sets the [`Outcome`](https://docs.rs/watchexec/latest/watchexec/action/enum.Outcome.html) of the Action. That's an intentionally restrained way to give commands to Watchexec's internal process manager, which is called [`Supervisor`](https://docs.rs/watchexec/latest/watchexec/command/struct.Supervisor.html).

What the supervisor does is it starts a process (technically it can supervise a _sequence_ of processes but let's keep things simple), and then holds a reference to it. When an action returns an Outcome, that may translate to a command to send to the Supervisor, like "send the process a signal" or "kill the process" or "wait for it to stop" or "start a new process" (which can only happen if the supervisor is idle aka doesn't have a running process at the moment). At the same time, the supervisor will also send an event down the events channel when the process quits, which are then handled in the exact same way by the whole machinery above.

When the supervisor is about to start a process, it will prepare it then call the `pre_spawn` handler, which can be used to modify the process call a bit or to send a notification out or print a message or something. After that returns, the supervisor starts the process, and then it calls the `post_spawn` handler. That one's less useful but has the new process's PID for example.

All these things are exposed, and if some things aren't I'm likely to accept PRs to expose them in the public API (unless I strongly feel they shouldn't be).

**Comment by emilHof** [comment]
thank you so much @passcod for all that detail! especially those drawings are very helpful for understanding this!

so the action worker could be seen as one big loop that processes incoming events and spawns a new `OutcomeWorker` for each process ? then the `OutcomeWorker` is responsible for propagating the `Outcome` to its `ProcessHolder` (which is the `Supervisor` wrapped in sync constructs)? i think you mentioned that a `Supervisor` can hold multiple commands, as `Supervisor::spawn()` feeds it a `Vec<Command>`, yet the `Command`s are executed in succession and one `Process` has to finish before the next starts?

that sounds awesome! it is really fun to be engaging with such an active and friendly maintainer as you !! :))

**Comment by passcod** [comment]
That's correct, yeah. The multiple commands stuff is done to support cargo-watch, though it's not used in the watchexec CLI.

**Comment by passcod** [comment]
Definitely have multiple supervisors, as that's its job, but the best thing to do would be to replicate/fork what the action::worker() is doing (I'm happy to merge PRs to break out and expose some parts, like the throttle+filter loop, so that it's easier) to support multiple supervisors managed by **one** action worker. Heck, that's the direction I'd want to go in eventually, so if you can manage a good implementation that does this I'd be happy to merge that directly.

**Comment by emilHof** [comment]
yay, that sounds amazing @passcod ! thank you for being so accommodating, it is making many things so much easier haha! i'll try and give it the best shot i have then, and get back to you, if that is alright hehe?

**Comment by passcod** [comment]
Sure, go for it. I'll be less responsive outside of weekends.

**Comment by emilHof** [comment]
yay, awesome!

one question I had for you and something i was wondering:

when assigning events to supervisors, after the initial filtering and throttling, so far there were two ways i have thought of to do this:

1. A new `RuleSet` trait with an `apply()` method. this maps `Vec<Event>` into `Vec<Category>` where `Category` is an enum with `Category::General(Event)` and `Category::Mapped(Event, Vec<Process>)` variants. From there the events can then be sent to the right processes.
2. Use the existing `Filter` and have one for each process. When a process is added to the `action::worker` it also needs to have its own `Filter`.
3. Some other option?

**Comment by passcod** [comment]
That's the beginning of the right direction I think. However making it a trait is not quite the pattern, I'd want a handler. In fact, why not change the `Action` handler itself, and have it gain a few methods: 

```rust
impl Action {
    // essentially route to all current processes,
	// aka the old behaviour for backcompat
	fn outcome(self, Outcome) {
		for id in self.current_processes() {
			self.on_process(id, Outcome, EventSet::All);
		}
	}

	fn on_process(&mut self, ProcessId, Outcome, EventSet);

	fn start_process(&mut self, Command, EventSet);

	// snapshotted at the time the Action is constructed, not a live query
	fn current_processes(&self) -> &[ProcessId];
}

enum EventSet {
	None,
	All,
	Some(Vec<EventId>),
}

impl Event {
	// impl as maybe a hash value for now
	fn id(&self) -> EventId;
}

/// opaque internal ID
struct EventId(u64);
```

**Comment by emilHof** [comment]
oh, yess, i really like that idea!

**Comment by passcod** [comment]
The outcome is one thing, but the process does get a list of events that triggered/affected it (in the prespawn handler), so the eventset is so that we can determine which events to give to that handler. A simpler implementation is to always give all the events to all processes' handlers, but that's a bit less helpful.

So for example, if you have `tree/foo` and `tree/bar`, and watchexec watches `tree`, but the action handler calls `on_process("for foo", Outcome::Restart)` only if files in `tree/foo/...` changed, it should probably also pass to that call only the events that match `tree/foo/...`, and not those that match e.g. `tree/bar/...`.

**Comment by emilHof** [comment]
now, looking over the code more closely, it may be that one of my prior questions has actually resolved itself haha!

would `on_process()` be called by the implementer of the `action_hanlder`? in that case, this makes a lot more sense to me now hehe!

would `Process`s or `ProcessId`s be something that is defined in the runtime data of the `Watchexec`? so user could add `Process`s at runtime?

**Comment by passcod** [comment]
> so user could add Processs at runtime?

that's the intent!

> would Processs or ProcessIds be something that is defined in the WorkingData of the Watchexec?

actually i forgot about this, but start_process() could return a ProcessId, so that the handler implementer can keep track of it.

(with these new methods, passing command to the working data becomes legacy use, essentially, with the new modern way being that the action handler always defines and starts the command instead!)

**Comment by emilHof** [comment]
ohh, okie!

so `Action` would also have a method like:

```rust
Action::start_process(&mut self, /* other params, such as command */) -> ProcessId;
```

And this method could be called an arbitrary amount by the implementer?

**Comment by passcod** [comment]
Yeah, sounds right.

**Comment by emilHof** [comment]
one more consideration here, that you will likely know how to answer much better than me:

if we let users start `Process`s directly in the `action_handler`, then how would users wanting to dynamically add additional `Process` be resolved? would the `action_handler` have access to some runtime state which tracks the running `Process`s and another state that is a queue of additional `Process` that are being requested to be started (maybe a `Receiver<ProcessReq>`)?

or is it maybe already the case that `action_handler` has access to the `RuntimeState` of the `Workexec`?

**Comment by emilHof** [comment]
thinking about this a little bit more, would a process something like this be appropriate:

1. `Event`s are collected, throttled, and filtered
2. Running processes are fetched from the last round of events and a new `Action` with `Arc<[Event]>` and `Arc<Mutex<Vec<Process>>>` is created. Maybe:
```rust
  Action::new(events: Arc<[Event]>, processes: Arc<Mutex<Vec<Process>>>) -> Action;
 ```
3. The` action_handler` is called to deal with determining `Outcome`s for all the `Process`s and start new `Process`.
4. Once the `action_handler` returns and does not `err` we now get a list of `Process`s instead of just one `Outcome`.
5. We save the processes for the next round.

Questions:
1. In step **3.**, if an outcome is not determined for a `Process` is this a no-op?
2. In step **3.**, how do we give users the ability to feed *new* `Process` dynamically?
3. Does the implementer calling `on_process(ProcessId, Outcome, EventSet)`  in step **3.** cause `Outcome::resolve(ProcessHolder)` to be called there and then, or should this still happen in the action worker between step **4.** and **5.**?

**Comment by passcod** [comment]
1. Yes. Can approximate to implying `Outcome::DoNothing` for the process, but more efficient to just not do anything to the process.
2. The action could be something more like:

```rust
struct Action {
	events: Arc<[Event]>,
	current_processes: Arc<[ProcessId]>,
	outcomes: Arc<Mutex<HashMap<ProcessId, (Resolution, EventSet)>>>,
}

impl Action {
fn start_process(&self, command, eventset) -> ProcessId {
	let id = ProcessId::new();
	self.outcomes.lock().insert(id,
		(Resolution::Start(command), eventset)
	);
	id
}

fn on_process(&self, process, outcome, eventset) {
	self.outcomes.lock().entry(process).replace_with( // i don't remember the entry API
		(Resolution::Start(outcome), eventset)
	);
}
}

impl ProcessId {
	fn new() -> Self {
		// random u64
	}
}

enum Resolution {
	Start(Command),
	Apply(Outcome),
}
```

3. The Action should be lightweight and "isolated" from the worker state. Only once we have it "back" from the handler, should we process its state (between steps 4 and 5). That also provides guarantees to the handler implementer about the execution order.

**Comment by emilHof** [comment]
hey, thank you for answering these so well haha! :D

As for **(1)** that makes a lot of sense! so if there is not `(Resolution, EventSet)` pair for an already existing `ProcessId`, we just leave that `Process`, if i am understanding this right ? In regard to **(3)**, i had not been thinking about the isolation part of the equation, and this makes a lot of sense now that you explain it so explicitly! :D

here's something else i now ran into.. 

1. should the `OutcomeWorker::spawn()` API and thus the internals be slightly altered, where we are no longer passing it `WorkingData` ?
2. would `pre_spawn_handler` and `post_spawn_handler` also be altered in their attributes and what the user gets to do with them? i.e. set `env`s for `Process` specifically, so it would maybe also have access to a new list of `ProcessId`s? 

also, as a quick aside, right `EventId` is generated via a `hash()` of the event and only happens once for an event, after which it keeps reusing the already generated `EventId`. `ProcessId` does it slightly differently and is using *xorshift\** to generate a pseudorandom `usize` value. 

would it be better if these both used the same method of id generation?

last consideration i was not so sure about and thought it may be good to check in with you about: if we give user's the ability to define a `ProcessId` in `start_process` they would be able to keep track of which `Process`s were already running through `Action::current_process()` and avoid creating new ones, instead just modifying the `Outcome`s for the already existing ones. would this make sense?

**Comment by passcod** [comment]
> if there is not (Resolution, EventSet) pair for an already existing ProcessId, we just leave that Process

yes

> OutcomeWorker::spawn() API and internals

As for internals, I don't really care, but I would prefer to avoid breaking the public API if that's at all possible.

> would pre_spawn_handler and post_spawn_handler also be altered in their attributes and what the user gets to do with them?

Again, I would prefer not to have breaking changes here. In fact I care more about changes here which is the "most oft used" API than the OutcomeWorker which is likely used less. If things can be achieved without breaking backward compat then I'm open :)

> would it be better if these both used the same method of id generation?

No, I think that those two have very different purposes, so that's expected.

---

I'm not sure if I quite understand your last paragraph, tbh. Can you rephrase it, sorry?

**Comment by emilHof** [comment]
> As for internals, I don't really care, but I would prefer to avoid breaking the public API if that's at all possible.

that is very fair hehe! is the `OutcomeWorker` in the public API or would you also be referring to API that is `pup(crate)`?

> If things can be achieved without breaking backward compat then I'm open :)

okie, awesome, roger that! back compat is the priority then ! :D a thought about that, would then prefer to keep the `RuntimeConfig` and `InitConfig` for backward compat? and spawn the `RuntimeConfig` defined `Command` if a user chooses to use this legacy option (as i think you may have called it)?

> No, I think that those two have very different purposes, so that's expected.

okie, awesome! that makes sense!

---

for the last paragraph, i was wondering if you think it would be good to let user's define `ProcessId`s themselves when creating a new `Process`, so they can refer to them on their own terms. for example, a user may choose to give each process a name which they use for the `ProcessId`. if this is not a great idea, for duplication reasons, or others, i definitely understand too!

**Comment by passcod** [comment]
Consider everything that's in the https://docs.rs/watchexec public API, and everything else fair game :)

**Comment by passcod** [comment]
> and spawn the RuntimeConfig defined Command if a user chooses to use this legacy option

definitely!

> in general, is adding additional methods to an already existing struct not considered to be breaking?

adding methods is fine. adding fields or variants is also fine as long as structs/enums are marked non-exhaustive. adding stuff to configs/workingdatas is explicitly allowed (but changing/removing isn't). as a special consideration, changing what a config builder method does is allowed (as long as it doesn't break reasonable prior usage).

**Comment by passcod** [comment]
> for the last paragraph, i was wondering if you think it would be good to let user's define ProcessIds themselves when creating a new Process, so they can refer to them on their own terms.

right, yeah. I don't think that's useful, I think if an implementer wants to do that, they should e.g. have a `hashmap<name, processid>`.

ProcessIds being opaque and small (should be `#[Copy]`) is a more useful value I think, and having eg strings allowed would break that. Also, might be a nice touch to make them contain a nonzero integer instead of a bare usize/u64 or whatever, so that niche optimisations can apply (ie `Option<ProcessId>` would be the same size as `ProcessId`).

**Comment by emilHof** [comment]
that is very good reasoning, you're right! the fact that user's can just store the mapping themselves and carry that cost is very true and does seem to make a lot more sense!

would unseeing something like `std::num::NonZeroUsize` suffice for getting the niche optimization?

**Comment by emilHof** [comment]
this may be an incorrect assumption, but would you prefer to not introduce an `id: EventId` field to `Event`  as to not break the public API?

this would maybe have the issue that, if `Event::id(&self) -> EventId` relies solely on hashing `Event` and its fields it may not always be the same hash value and thus the `EventId` might differ from two calls to the same `Event`. or is this a wrong understanding that i am having? that could very well be the case hehe!

**Comment by emilHof** [comment]
and one last question for tonight, i promise hehe!

using `Arc<Mutex<HashMap<ProcessId, (Resolution, EventSet)>>>` has the side effect of turning `Action::outcome()` and `Action::on_process()` into `async fn`s, at least of `tokio::mutex` is what is being used for the `Mutex<T>`. this seems to be a major breaking change, at least to the `Action` API... to work around this action now uses `crossbeam::AtomicCell<T>` instead of `tokio::sync::Mutex<T>`, which comes at the price of swapping the `HashMap` out of and into the cell each time `Action::on_process()` is called.. though this way it need to be `async` yay! :D

is this an okay way of doing it, or would you prefer another approach?

**Comment by passcod** [comment]
> this may be an incorrect assumption, but would you prefer to not introduce an id: EventId field to Event as to not break the public API?

that's right

> this would maybe have the issue that, if Event::id(&self) -> EventId relies solely on hashing Event and its fields it may not always be the same hash value and thus the EventId might differ from two calls to the same Event.

hmm, yeah. I was trying to make it cheap but actually that's probably more expensive anyway. Let's forget about EventIds and just use straight Events. If there's performance issues due to that in the future I'll figure out the best thing and break the API at that time, but let's keep it simple for here.

(I think that will mean that technically Action handlers could return brand new synthetic events as well as those that originally came in, but I can't really see a usecase for that at the moment, so.)

**Comment by passcod** [comment]
> would using something like std::num::NonZeroUsize suffice for getting the niche optimization?

Yes, any nonzero. I think I slightly prefer using a fixed-sized one, so u64/NonZeroU64 rather than usize, but that's not terribly important.

**Comment by passcod** [comment]
> is this an okay way of doing it

yeah, and as long as that remains opaque to the handler (as it's only ever used via the methods) it can be swapped later on with whatever.

---

wait, no, hold on

Right so [tokio's Mutex can do a blocking lock](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.blocking_lock) so `Action::outcome()` could be implemented in terms of that, and the new methods async instead. If that makes types too complicated fallback to the crossbeam thing but let's try that way first.

**Comment by emilHof** [comment]
> Let's forget about EventIds and just use straight Events.

okie, that makes sense! saving on the direct comparison would definitely have been nice, but you'll very likely know much better how to apply changes that might be *breaking*, so leaving that to you sounds good ! :))

just straight events for now it is then ehehe! :D

> u64/NonZeroU64 rather than usize

okie! is there a reason, i am sure there is a very good one haha, for using fixed sized types in this context ?

> implemented in terms of that, and the new methods async instead

ohh, i was not aware of this! that is sensible to do it that way then yess! one thing though, `Action::outcome()` just calls `Action::on_process()` repeatedly with all the `ProcessId`s, so if `Action::on_process()` is `async`, would it be hard to make `Action::outcome()` `sync`?

**Comment by passcod** [comment]
> is there a reason

it's more vibes really, but:
- it's easier to reason about what size it is if the size is explicit, and eg in this case we're generating it from xorshift, right, so we'd be specifically using xorshift64, not xorshift32 or something.
- usize is used most for pointer-sized things and indexes/capacities/lengths, that kind of thing, and I think reserving it in a way for those is useful, so that if you see a usize you know that it's probably used for one of those things.
- also of course on 32-bit archs usize and isize are 32-bit long, while a u64 is always 8 bytes.

**Comment by emilHof** [comment]
> it's more vibes really

haha, i love that so much !! :D 

those are really good reasons, i hadn't really considered any of them! especially the xorshift has great potential in causing issues! will definitely make those changes; now there's no way around it hehe!

okie, that makes sense! right now, the way i ended up working around it was with an internal helper:

```rust
fn on_process_with_guard(guard: Mutex<...>, /* other args */)
```

that is then used by both:

```rust
pub async fn on_process_with(/* other args */) {
    let guard = /* get guard with await */;
    self.on_process_with_guard(guard, /* ... */);
}
```

is this a bit too contrived and should i just go with your solution? yours does have the advantage of avoiding a helper with a messy contract haha!

**Comment by passcod** [comment]
Oh, cool. All of this is fine as you've done it. Good choices :)

**Comment by emilHof** [comment]
yay, awesome!! well, under amazing instruction hehe!

hopefully i am not bothering you an incredible amount with all of these constant questions! definitely do not feel obliged to respond ask quickly as you are doing! it is truly remarkable how active of a maintainer you are !!

one quick question, hehe, regarding the `pre_spawn_handler` and `PreSpawn`. in order to give user the ability to determine which process the `pre_spawn_handler` is being called for, would it be a good idea to provide them with the `ProcessId` through `PreSpawn`? this may have the downside of passing the `ProcessId` all the way to `Supervisor` where `PreSpawn::new()` gets called and thus would imply possible changes to the signature of `OutcomeWorker::spawn()` and `Supervisor::spawn()` which would get passed the `ProcessId`.. do you think there is maybe a better way of doing this, attaching the `ProcessId` to the `ProcessHolder`, or some other option?

**Comment by passcod** [comment]
Supervisor definitely should have the ProcessId.

Looking at `spawn()`, though, we've got a pretty clear refactor opportunity/need here, because that function/constructor takes absolutely too many arguments, and adding arguments here at a pretty core point of expansion is probably going to need to happen again.

So: that should be a builder!

```rust
#[non_exhaustive]
pub struct SupervisorBuilder {
	// ...all the args to spawn() as fields
	// plus the new process_id
}

impl SupervisorBuilder {
	pub fn build(self) -> Supervisor { ... }
}
```

That leaves the question of what to do with the existing `spawn()`. At this stage, I think it should make a new random process ID, its documentation should say that's what it's doing, and it should be marked deprecated. That way, any existing usage will keep working (and devs will get a deprecated warning), even though the watchexec internals won't use it anymore.

Then, at some later point in the future, `spawn()` will be removed as a breaking change!

---

(Also I'm noticing that `grouped: bool` and the vec of commands next to each other and I think that really, I should have made `grouped` part of the `Command` type, but that's an unrelated breaking change I'll need to do later.)

**Comment by passcod** [comment]
Also thinking about terminology, maybe `ProcessId` should be renamed `SupervisorId` or `CommandId`, so that it's not ambiguous/confused with OS-level process IDs, and so that it doesn't imply that each new process spawn (like within a supervisor) gets a new `ProcessId`.

**Comment by passcod** [comment]
> instantiate it as a literal

yes, just as a literal to start with, as most of those fields are required anyway. Methods can be added later quite easily.

> how grouped is used

`grouped` controls whether a process group is used, which makes it such that if a command starts its own subprocesses, those are reaped at the same time as the main command. That's generally what's wanted, the `grouped` option is there mostly so it can be disabled if the user wants more control for whatever reason.

(In the future I want to add even more "process grouping related" options, such as running in a pty or even running in a namespace (think Docker).)

> this being the actual OS-level PID?

that's right, yeah

**Comment by emilHof** [comment]
> That's generally what's wanted, the grouped option is there mostly so it can be disabled

ohh, okie! the namespacing sounds super interesting! how would this work and shape out?

**Comment by passcod** [comment]
not sure yet. presumably, make a [cgroup](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html) and start the process within it. but one thing at a time!

**Comment by emilHof** [comment]
that sounds interesting! my knowledge of this, especially `cgroup` is super limited, but i'll definitely give this a look now that you mention it! 

hehe, yess, one thing a time does make sense! :D

also, would it make sense to start opening a *WIP* PR to maybe make it easier for you to view what you'd still like to see added/changed?

**Comment by passcod** [comment]
Yes please :)

**Comment by emilHof** [comment]
okie, will do :vulcan_salute: ! :D
